subsys/bluetooth/audio/has_client.c

```diff
@@ -895,108 +1413,191 @@ int bt_has_client_presets_read(struct bt_has *has, uint8_t start_index, uint8_t
 
-	if (!inst->conn) {
+	CHECKIF(inst == NULL) {
 		return -ENOTCONN;
 	}
 
-	if (sync && !bt_has_feat_is_preset_oob_sync_supported(inst->has.features)) {
-		return -EOPNOTSUPP;
+	if (inst->state != STATE_CONNECTED) {
+		return -ENOTCONN;
 	}
 
-	if (atomic_test_bit(inst->flags, HAS_DISCOVER_IN_PROGRESS) ||
-	    atomic_test_and_set_bit(inst->flags, HAS_CP_OPERATION_IN_PROGRESS)) {
-		return -EBUSY;
+	if (sync && !bt_has_features_check_preset_sync_supp(inst->features)) {
+		return -EOPNOTSUPP;
 	}
 
 	opcode = sync ? BT_HAS_OP_SET_PREV_PRESET_SYNC : BT_HAS_OP_SET_PREV_PRESET;
 
 	return preset_set_next_or_prev(inst, opcode);
 }
 
+int bt_has_client_info_get(const struct bt_has_client *inst, struct bt_has_client_info *info)
+{
+	CHECKIF(inst == NULL) {
+		LOG_ERR("NULL inst");
+
+		return -EINVAL;
+	}
+
+	CHECKIF(info == NULL) {
+		LOG_ERR("NULL info");
+
+		return -EINVAL;
+	}
+
+	info->addr = &inst->addr;
+	info->id = inst->id;
+	info->type = bt_has_features_get_type(inst->features);
+	info->features = inst->features;
+	info->caps = get_capabilities(inst);
+	info->active_index = inst->active_index;
+
+	return 0;
+}
+
+int bt_has_client_conn_get(const struct bt_has_client *inst, struct bt_conn **conn)
+{
+	CHECKIF(inst == NULL) {
+		LOG_ERR("NULL inst");
+
+		return -EINVAL;
+	}
+
+	CHECKIF(conn == NULL) {
+		LOG_ERR("NULL conn");
+
+		return -EINVAL;
+	}
+
+	if (inst->conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	*conn = inst->conn;
+
+	return 0;
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	struct bt_conn_info info = { 0 };
+	struct bt_has_client *inst;
+
+	LOG_DBG("conn %p err %d", (void *)conn, err);
+
+	if (err != 0) {
+		return;
+	}
+
+	err = bt_conn_get_info(conn, &info);
+	__ASSERT_NO_MSG(err == 0);
+
+	inst = inst_find_by_addr(info.le.dst);
+	if (inst == NULL) {
+		return;
+	}
+
+	__ASSERT_NO_MSG(inst->conn == NULL);
+	inst->conn = bt_conn_ref(conn);
+
+	inst->err = inst_connect(inst);
+	if (inst->err != 0) {
+		LOG_ERR("Failed to connect profile (err %d)", inst->err);
+	}
+}
+
 static void disconnected(struct bt_conn *conn, uint8_t reason)
 {
-	struct has_inst *inst = inst_by_conn(conn);
+	struct bt_has_client *inst;
 
-	if (!inst) {
+	LOG_DBG("conn %p reason %d", (void *)conn, reason);
+
+	inst = inst_find_by_conn(conn);
+	if (inst == NULL) {
 		return;
 	}
 
-	if (atomic_test_bit(inst->flags, HAS_DISCOVER_IN_PROGRESS)) {
-		discover_failed(conn, -ECONNABORTED);
+	if (inst->state == STATE_CONNECTING ||
+	    inst->state == STATE_RECONNECTING) {
+		inst->err = -ECONNABORTED;
 	}
 
-	inst_cleanup(inst);
+	inst_state_set(inst, STATE_DISCONNECTED);
+}
+
+static void identity_resolved(struct bt_conn *conn, const bt_addr_le_t *rpa,
+			      const bt_addr_le_t *identity)
+{
+	struct bt_has_client *inst;
+
+	LOG_DBG("conn %p %s -> %s", (void *)conn, bt_addr_le_str(rpa), bt_addr_le_str(identity));
+
+	inst = inst_find_by_conn(conn);
+	if (inst != NULL) {
+		bt_addr_le_copy(&inst->addr, identity);
+	}
```

Comment:
Does this need to be thread-safe?

Reply:
I am not sure. Can you elaborate, i.e why this particular operation? The callback is ran from cooperative thread, and I think there is assumption that the Bluetooth API in general has to be called from cooperative thread as well.

Reply:
`bt_has_client_bind` calls `inst_alloc` which also copies an address to `&inst->addr`, so we have a public API function that modifies the same value as this callback does, with a non-atomic operation. 

It's very unlikely to be an issue, so it's OK as is I think